### PR TITLE
Alison/transactions

### PR DESCRIPTION
**Pull request by adowski**
## Transactions Page
Created `ArtistTransactions` page, `CreateTransaction` form, `Modal` class

### Related PRs
Will add error messages after toastr gets merged!!

### Migrations
Created `transactions`
Added belongs_to relationship w `artists`, `buyers`, `works`
Added `comment` to transactions
Changed `price` type to decimal

### Tests Performed, Edge Cases
Currently artists cannot actually record a transaction until the necessary fields are filled

### Screenshots
image

### Also

- I attempted to learn how to style lol but didn't really know what I was doing so I can fix/delete that
- Currently all transactions are made between the current logged in artist, Bill Gates, and the Mona Lisa work, eventually the buyer/work props passed in will be the ones associated with the request

CC: @clairelin135

### CODE REVIEW

File: app/javascript/components/Modal.jsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import PropTypes from "prop-types";
+import React from "react";
+
+class Modal extends React.Component {
```
Comment (clairelin135) [inline app/javascript/components/Modal.jsx]: Sorry! In addition, you should also use an existing modal library. This is because the way you implement it currently, you can still access elements on the screen behind the modal. Instead, we need to prevent the user from making these changes. You can implement an existing modal library such as react-modal or ReactBootstrap.Modal

File: app/views/transactions/new.html.slim [hunk 2]
```diff
@@ -0,0 +1,4 @@
+= render "partials/header"
```
Comment (clairelin135) [inline app/views/transactions/new.html.slim]: You don't actually need to render the header since its been imported in application.html.erb, meaning that by default it is included in every view file

File: app/views/transactions/show.html.slim [hunk 3]
```diff
@@ -0,0 +1,4 @@
+= render "partials/header"
```
Comment (clairelin135) [inline app/views/transactions/show.html.slim]: Same thing

File: app/assets/javascripts/helpers/api_routes.js [hunk 4]
```diff
@@ -14,13 +14,14 @@ class ApiRoutes {
...
   get artists() {
     return {
-      index: `/api/artists`,
-      show: id => `/api/artists/${id}`,
-      update: id => `/api/artists/${id}`,
-      delete: id => `/api/artists/${id}`,
-      works: id => `/api/artists/works/${id}`,
-      requests: id => `/api/artists/requests/${id}`
-    };
+      show        : (id) => `/api/artists/${id}`,
```
Comment (clairelin135) [inline app/assets/javascripts/helpers/api_routes.js]: I think ur missing the
index: `api/artists`
route

Comment (clairelin135) [reply app/assets/javascripts/helpers/api_routes.js]: you don't need the AllArtists route below since you're retrieving the route from `APIRoutes.artists.index`

File: app/controllers/api/artists_controller.rb [hunk 5]
```diff
@@ -53,4 +53,11 @@ def requests
     end
   end
 
+  def transactions
+    artist = Artist.find(params[:id])
+    transactions = artist.transactions
+    render json: transactions,
+        each_serializer: TransactionSerializer
```
Comment (clairelin135) [inline app/controllers/api/artists_controller.rb]: I think that you didn't upload your transactionserializer class

File: config/routes.rb [hunk 6]
```diff
@@ -20,18 +20,25 @@
   get '/artists/:id' => 'artists#show', as: :artistid
   get '/artists/', to: 'artists#all_artists'
 
+  get '/transactions/' => 'transactions#show'
```
Comment (clairelin135) [inline config/routes.rb]: Typically when you want to retrieve all entries in a db, it is an index method so we should rename this to transactions#index. show is normally used for a single entry

Also, can you add a link to the transactions within the header so that upon hovering over artist name you can click on transactions? You should probably be able to pattern-match based on what's already there but if you have q's you should ask me

File: app/javascript/components/CreateTransaction.jsx [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+  return (
+      <div className="create-transaction-form">
+        <h2 className="f4 lh-title submit">Submit Details</h2>
+
+        <p className="f6 lh-copy price">Price  </p>
+        <input
+          type="TEXT"
+          name="price"
+          id="price"
+          value={this.state.transaction.price}
```
Comment (clairelin135) [inline app/javascript/components/CreateTransaction.jsx]: Bump! We should be styling the price within the component so that there isn't room for error for the user to input values. You can do something like `value={this.currencyDisplay(this.state.transaction.price)}` so that the price is automatically displayed formatted correctly

Comment (clairelin135) [reply app/javascript/components/CreateTransaction.jsx]: you should be able to find the function online somewhere and you can just copy paste

File: app/javascript/components/CreateTransaction.jsx [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+        />
+
+        <p>{this.renderErrorMessages("price")}</p>
+
+        <div className="f6 lh-copy drop-down">Type
+          <select name="transaction_type"
+                  value={this.state.transaction.transaction_type}
+                  onChange={this.handleChange}>
+              <option value="choose">choose a type</option>
+              {  Object.keys(this.state.types).map((obj) => { return <option>{obj}</option> }) }
```
Comment (clairelin135) [inline app/javascript/components/CreateTransaction.jsx]: Looks like we're getting a `keyError` value in the chrome console here. You should instead write something like this:
```
Object.keys(this.state.types).map((obj, i) => {return <option key={i}>{obj}</option>})
```
which would essentially input a unique key prop based on the index of the array

File: app/javascript/components/CreateTransaction.jsx [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+          id="price"
+          value={this.state.transaction.price}
+          onChange = {this.handleChange}
+        />
+
+        <p>{this.renderErrorMessages("price")}</p>
+
+        <div className="f6 lh-copy drop-down">Type
+          <select name="transaction_type"
+                  value={this.state.transaction.transaction_type}
```
Comment (clairelin135) [inline app/javascript/components/CreateTransaction.jsx]: We're also getting a lot of stylistic console errors while rendering the form. Can we fix all of these?

File: app/javascript/components/ArtistTransactions.jsx [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+              <span className="i"> {transaction.buyer_name} </span>
+              <span className="i">{transaction.transaction_type}</span>
+            </p>
+            <p>
+              <span>{transaction.comment}</span>
+            </p>
+          </div>
+        ))}
+
+        <ReactModal className="ba pa3 modal" isOpen={this.state.show} onRequestClose={this.hideModal}>
```
Comment (clairelin135) [inline app/javascript/components/ArtistTransactions.jsx]: I like the way you structured this. However, I have a couple of suggestions:

- All modals contain shared options such as the submit and cancel buttons, as well as shared styling. We should create a separate `StyledModal` class that simply accepts the contents through child components to create another level of abstraction, with submit/cancel buttons & props. From there, additional modal components can extend the `StyledModal` class
- It seems extraneous to store in the state whether the modal is closed or open. Truly, the parent component doesn't need to know whether the modal is closed or open. Instead, we should store this state within the `StyledModal` class.
- A modal component holistically is a button that upon clicked displays the actual modal. Thus, you can render the button within the `StyledModal` class and use that to manage displaying the modal. Then, you can import the component, and it will simply display as a button normally

Comment (clairelin135) [reply app/javascript/components/ArtistTransactions.jsx]: Also, this ensures modularity! Then we can extend the class for other uses

Comment (clairelin135) [reply app/javascript/components/ArtistTransactions.jsx]: A good example to look at would be SJC's

File: app/javascript/components/CreateTransaction.jsx [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+          />
+
+          <p>{this.renderErrorMessages("purchase_date")}</p>
+        </div>
+      )
+    }
+  }
+
+  renderRentalDates = () => {
+    if (this.state.transaction.transaction_type == "rental") {
```
Comment (clairelin135) [inline app/javascript/components/CreateTransaction.jsx]: You also should be able to input a purchase date for rentals too

File: app/javascript/components/CreateTransaction.jsx [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+    const valid = this.state.formValid;
+    const payload = {
+      "type": this.state.transaction.transaction_type,
+      "start_date": this.state.transaction.start_date,
+      "end_date": this.state.transaction.end_date,
+      "purchase_date": this.state.transaction.purchase_date,
+      "price": this.state.transaction.price,
+      "buyer_id": 0,
+      "artist_id": artist_id,
+      "work_id": 0,
```
Comment (clairelin135) [inline app/javascript/components/CreateTransaction.jsx]: let's make sure to fix this before merging!

File: app/javascript/components/CreateTransaction.jsx [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+        fields.type_valid = value != null && value != 'choose';
+        errors.transaction_type = fields.type_valid ? '' : 'Please specify a type of transaction.'
+        break;
+      case 'purchase_date':
+        fields.purchase_valid = value != null && this.state.transaction.transaction_type == "purchase";
+        errors.purchase_date = fields.purchase_valid ? '' : errors.purchase_date
+        break;
+      case 'start_date':
+        fields.start_valid = value != null && this.state.transaction.transaction_type == "rental";
+        errors.start_date = fields.start_valid ? '' : errors.start_date
```
Comment (clairelin135) [inline app/javascript/components/CreateTransaction.jsx]: this stuff is quite a pain to manage, looks fine to me tbh but we may need to refactor down the line
for now, this looks good to me
